api/kroxylicious-filter-api/src/main/templates/Kproxy/SpecificFilterArrayInvoker.ftl

```diff
@@ -0,0 +1,157 @@
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ${outputPackage};
+
+import org.apache.kafka.common.message.ApiMessageType;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.message.ResponseHeaderData;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.ApiMessage;
+
+/**
+ * Invoker for KrpcFilters that implement any number of Specific Message interfaces (for
+ * example {@link io.kroxylicious.proxy.filter.AlterConfigsResponseFilter}.
+ */
+class SpecificFilterArrayInvoker implements FilterInvoker {
+
+    private final FilterInvoker[] requestInvokers;
+    private final FilterInvoker[] responseInvokers;
+
+    SpecificFilterArrayInvoker(KrpcFilter filter) {
+        this.requestInvokers = new FilterInvoker[ApiMessageType.values().length];
+        this.responseInvokers = new FilterInvoker[ApiMessageType.values().length];
+        <#list messageSpecs as messageSpec>
+        <#if messageSpec.type?lower_case == 'request'>
+        this.requestInvokers[${messageSpec.apiKey.get()}] = filter instanceof ${messageSpec.name}Filter ? new ${messageSpec.name}FilterInvoker((${messageSpec.name}Filter) filter) : null;
+        </#if>
+        </#list>
+        <#list messageSpecs as messageSpec>
+        <#if messageSpec.type?lower_case == 'response'>
+        this.responseInvokers[${messageSpec.apiKey.get()}] = filter instanceof ${messageSpec.name}Filter ? new ${messageSpec.name}FilterInvoker((${messageSpec.name}Filter) filter) : null;
+        </#if>
+        </#list>
+    }
+
+    /**
+     * Apply the filter to the given {@code header} and {@code body} using the given {@code filterContext}.
+     * @param apiKey The request api key.
+     * @param apiVersion The request api version.
+     * @param header The request header.
+     * @param body The request body.
+     * @param filterContext The filter context.
+     */
+    @Override
+    public void onRequest(ApiKeys apiKey,
+                           short apiVersion,
+                           RequestHeaderData header,
+                           ApiMessage body,
+                           KrpcFilterContext filterContext) {
+        FilterInvoker invoker = requestInvokers[apiKey.id];
```

Comment:
We should check if apiKey.id is going to go out of bounds in the array, this is possible if we are running against a newer version of kafka-clients than this was built against.